### PR TITLE
Allow us to dispose batches asynchronously

### PR DESCRIPTION
**Pull request by jasonmalinowski**
This reduces contention on the Workspace lock, where we often have a bunch of threads waiting while one project completes it's batch.

**THIS PR IS CURRENTLY A HACK TO BENCHMARK THIS. PLEASE IGNORE.**

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs [hunk 1]
```diff
@@ -1648,14 +1688,17 @@ public void ReorderFiles(ImmutableArray<string> filePaths)
                     }
                     else
                     {
-                        _project._workspace.ApplyBatchChangeToWorkspace(solution =>
+                        // If we're passing useAsync: false, we should always get a task that was already completed.
```
Comment (ryzngard) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:1691]: Is that true? There are other awaits in the method as well.

Comment (ryzngard) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs:1691]: Ah nm... still seems suspect though. Why don't we split into 3 methods? 

1. Sync
2. Async

3. Sync work both async and sync point to

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs [hunk 2]
```diff
@@ -76,6 +77,7 @@ internal interface IWorkspaceProjectContext : IDisposable
 
         void StartBatch();
         void EndBatch();
+        ValueTask EndBatchAsync();
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:80]: 1. do we want both this and the Sync version?
2. do we need StartBatchAsync?

File: src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs [hunk 3]
```diff
@@ -252,12 +253,19 @@ public void SetRuleSetFile(string filePath)
         public void StartBatch()
             => _batchScopes.Enqueue(_visualStudioProject.CreateBatchScope());
 
+        [Obsolete($"Use {nameof(EndBatchAsync)}.")]
         public void EndBatch()
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProject_IWorkspaceProjectContext.cs:257]: 1. make an explicit impl?
2. make interface method obsolete as well?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs:80]: We don't want the sync version but yes forgot to make it obsolete.

For StartBatchAsync, at this point I don't think so. The locking design here for the workspace is each project has it's own lock, and then there's also the big workspace lock when we're actually making changes. The intent with a batch is each of the operations within the batch don't require the big workspace lock until you actually apply it -- hence the EndBatchAsync is being made async here to avoid contention on the big global lock. StartBatch only is taking the project lock, which is unlikely to have any contention unless the project system is updating a single project from multiple threads at once.

tl;dr: the idea of batching is everything is cheap until the EndBatch, so we really shouldn't need it.
